File: kotlin-node/karakum/src/jsMain/resources/nameResolvers/resolveChildProcessOptionsName.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default (node, context) => {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (!sourceFileName.endsWith("child_process.d.ts")) return null

    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    let parameter = getParent(node)
    if (ts.isParenthesizedTypeNode(parameter)) {
        parameter = getParent(parameter)
    }
    if (ts.isUnionTypeNode(parameter)) {
        parameter = getParent(parameter)
    }

    if (!parameter) return null
    if (!ts.isParameter(parameter)) return null
    if (!ts.isIdentifier(parameter.name)) return null

    const parameterName = parameter.name.text

    const signatureDeclaration = getParent(parameter)
    if (!signatureDeclaration) return null
    if (!ts.isFunctionDeclaration(signatureDeclaration)) return null
    if (signatureDeclaration.name === undefined) return null

    const parentName = signatureDeclaration.name.text

    if (
        parentName === "exec"
    ) {
        let infix = "String"

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[0]
            && ts.isTypeLiteralNode(node.types[0])
            && node.types[0].members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "encoding"
                && ts.isUnionTypeNode(member.type)
                && member.type.types[0]
                && ts.isLiteralTypeNode(member.type.types[0])
                && ts.isStringLiteral(member.type.types[0].literal)
                && member.type.types[0].literal.text === "buffer"
            ))
        ) {
            infix = "Buffer"
        }

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[0]
            && ts.isTypeReferenceNode(node.types[0])
            && ts.isIdentifier(node.types[0].typeName)
            && node.types[0].typeName.text === "ObjectEncodingOptions"
        ) {
            infix = "ObjectEncoding"
        }

        return `${karakum.capitalize(parentName)}${infix}${karakum.capitalize(parameterName)}`
    }

    if (
        parentName === "execFile"
    ) {
        let infix = ""

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[0]
            && ts.isTypeReferenceNode(node.types[0])
            && ts.isIdentifier(node.types[0].typeName)
            && node.types[0].typeName.text === "ObjectEncodingOptions"
        ) {
            infix = "ObjectEncoding"
        }

        return `${karakum.capitalize(parentName)}${infix}${karakum.capitalize(parameterName)}`
    }

    return `${karakum.capitalize(parentName)}${karakum.capitalize(parameterName)}`
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/nameResolvers/resolveChildProcessOptionsName.kt
package node.karakum.nameResolvers

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import node.karakum.util.nullable
import typescript.Node
import typescript.asArray
import typescript.isFunctionDeclaration
import typescript.isIdentifier
import typescript.isIntersectionTypeNode
import typescript.isLiteralTypeNode
import typescript.isParameter
import typescript.isParenthesizedTypeNode
import typescript.isPropertySignature
import typescript.isStringLiteral
import typescript.isTypeLiteralNode
import typescript.isTypeReferenceNode
import typescript.isUnionTypeNode

fun resolveChildProcessOptionsName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

    val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
    ensure(sourceFileName.endsWith("child_process.d.ts"))

    val parameter = ensureNotNull(
        typeScriptService.getParent(node)
            ?.let {
                if (isParenthesizedTypeNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
            ?.let {
                if (isUnionTypeNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
    )
    ensure(isParameter(parameter))

    val parameterNameNode = parameter.name
    ensure(isIdentifier(parameterNameNode))
    val parameterName = parameterNameNode.text

    val signatureDeclaration = ensureNotNull(typeScriptService.getParent(parameter))
    ensure(isFunctionDeclaration(signatureDeclaration))

    val signatureName = ensureNotNull(signatureDeclaration.name)
    val parentName = signatureName.text

    nullable {
        ensure(parentName == "exec")

        nullable {
            ensure(isIntersectionTypeNode(node))

            val firstType = ensureNotNull(node.types.asArray().first())
            ensure(isTypeLiteralNode(firstType))
            ensure(firstType.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "encoding")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isUnionTypeNode(propertyType))

                    val firstPropertyType = ensureNotNull(propertyType.types.asArray().first())
                    ensure(isLiteralTypeNode(firstPropertyType))

                    val firstPropertyTypeLiteral = firstPropertyType.literal
                    ensure(isStringLiteral(firstPropertyTypeLiteral))
                    ensure(firstPropertyTypeLiteral.text == "buffer")
                } != null
            })

            "${parentName.replaceFirstChar { it.titlecase() }}Buffer${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: nullable {
            ensure(isIntersectionTypeNode(node))

            val firstType = ensureNotNull(node.types.asArray().first())
            ensure(isTypeReferenceNode(firstType))

            val typeName = firstType.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "ObjectEncodingOptions")

            "${parentName.replaceFirstChar { it.titlecase() }}ObjectEncoding${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: run {
            "${parentName.replaceFirstChar { it.titlecase() }}String${parameterName.replaceFirstChar { it.titlecase() }}"
        }
    } ?: nullable {
        ensure(parentName == "execFile")

        ensure(isIntersectionTypeNode(node))

        val firstType = ensureNotNull(node.types.asArray().first())
        ensure(isTypeReferenceNode(firstType))

        val typeName = firstType.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "ObjectEncodingOptions")

        "${parentName.replaceFirstChar { it.titlecase() }}ObjectEncoding${parameterName.replaceFirstChar { it.titlecase() }}"
    } ?: run {
        "${parentName.replaceFirstChar { it.titlecase() }}${parameterName.replaceFirstChar { it.titlecase() }}"
    }
}
